src/test_memcachedsolution.cpp

```diff
@@ -0,0 +1,864 @@
+  std::string data_in = "SimpleMemcachedSolutionTest.AddGet";
+  std::string data_out;
+
+  rc = this->set_data(data_in, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out, data_in);
+}
+
+/// Add two keys and retrieve them.
+TYPED_TEST(SimpleMemcachedSolutionTest, AddGetTwoKeys)
+{
+  uint64_t cas1 = 0;
+  uint64_t cas2 = 0;
+  Store::Status rc;
+  std::string key1 = this->_key;
+  this->get_new_key();
+  std::string key2 = this->_key;
+  std::string data_in1 = "SimpleMemcachedSolutionTest.AddGetTwoKeys1";
+  std::string data_in2 = "SimpleMemcachedSolutionTest.AddGetTwoKeys2";
+  std::string data_out1;
+  std::string data_out2;
+
+  rc = this->set_data(key1, data_in1, cas1);
+  EXPECT_EQ(Store::Status::OK, rc);
+  rc = this->set_data(key2, data_in2, cas2);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  rc = this->get_data(key1, data_out1, cas1);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out1, data_in1);
+  rc = this->get_data(key2, data_out2, cas2);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out2, data_in2);
+}
+
+/// Add a key that expires.
+TYPED_TEST(SimpleMemcachedSolutionTest, AddGetExpire)
+{
+  uint64_t cas = 0;
+  Store::Status rc;
+  std::string data_in = "SimpleMemcachedSolutionTest.AddGetExpire";
+  std::string data_out;
+
+  rc = this->set_data(data_in, cas, 1);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out, data_in);
+
+  sleep(2);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::NOT_FOUND, rc);
+}
+
+/// Add a key, retrieve it and try to update it twice. The second attempt fails
+/// due to data contention.
+TYPED_TEST(SimpleMemcachedSolutionTest, AddSetSetDataContentionSet)
+{
+  uint64_t cas = 0;
+  Store::Status rc;
+  std::string data_in = "SimpleMemcachedSolutionTest.AddSetSetDataContentionSet";
+  std::string data_out;
+
+  rc = this->set_data(data_in, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out, data_in);
+
+  data_in = "SimpleMemcachedSolutionTest.AddSetSetDataContentionSet_New1";
+  rc = this->set_data(data_in, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  std::string failed_data_in = "FAIL";
+  rc = this->set_data(data_in, cas);
+  EXPECT_EQ(Store::Status::DATA_CONTENTION, rc);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out, data_in);
+
+  data_in = "SimpleMemcachedSolutionTest.AddSetSetDataContentionSet_New2";
+  rc = this->set_data(data_in, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out, data_in);
+}
+
+/// Add a key and retrieve it. Try to update the key with a new value and also
+/// to update it with an expiry of 0. The delete update fails due to data
```

Comment:
The API to the store distinguishes between a delete and immediate expiry, so this comment shouldn't mention deletes.